ddl/job_table.go

```diff
@@ -0,0 +1,612 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/logutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	addingDDLJobGeneral = "/tidb/ddl/add_ddl_job_general"
+	addingDDLJobReorg   = "/tidb/ddl/add_ddl_job_reorg"
+)
+
+func (dc *ddlCtx) insertRunningDDLJobMap(id int64) {
+	dc.runningJobs.Lock()
+	defer dc.runningJobs.Unlock()
+	dc.runningJobs.runningJobMap[id] = struct{}{}
+}
+
+func (dc *ddlCtx) deleteRunningDDLJobMap(id int64) {
+	dc.runningJobs.Lock()
+	defer dc.runningJobs.Unlock()
+	delete(dc.runningJobs.runningJobMap, id)
+}
+
+func (dc *ddlCtx) getRunningDDLJobIDs() string {
+	dc.resetRunningIDs()
```

Comment:
It seems that we need to lock `runningJobIDs`() too, otherwise, this will have a data race.

Reply:
@zimulala Done, PTAL